command-line-interface.md

```diff
@@ -0,0 +1,234 @@
+        * *stdout:* The runtime MUST attach its stdout directly to the container process without writing to it.
+        * *stderr:* The runtime MUST attach its stderr to the container process, and MUST not write to it unless it exits with a non-zero code.
+* *Environment variables*
+    * *`LISTEN_FDS`:* The number of file descriptors passed.
+      For example, `LISTEN_FDS=2` would mean that the runtime MUST pass file descriptors 3 and 4 to the container process (in addition to the standard streams) to support [socket activation][systemd-listen-fds].
+* *Exit code:* Zero if the container was successfully created and non-zero on errors.
+
+Callers MAY block on this command's successful exit to trigger post-create activity.
+
+#### Console socket
+
+The [`AF_UNIX`][unix-socket] [`SOCK_SEQPACKET`][socket-types] used by [`--console-socket`](#create) handles request and response messages between a client and server.
+The server MUST send a single response for each client request.
+The [normal data][socket-queue] ([`msghdr.msg_iov*`][socket.h]) of all messages MUST be [UTF-8][] [JSON](glossary.md#json).
+
+##### Requests
+
+All requests MUST contain a **`type`** property whose value MUST one of the following strings:
+
+* `terminal`, if the request is passing a [pseudoterminal master][posix_openpt.3].
+    When `type` is `terminal`, the request MUST also contain the following properties:
+
+    * **`container`** (string, REQUIRED) The container ID, as set by [create](#create).
+
+    The message's [ancillary data][socket-queue] (`msg_control*`) MUST contain a single [`cmsghdr`][socket.h]) with:
+
+    * `cmsg_type` set to [`SOL_SOCKET`][socket.h],
+    * `cmsg_level` set to [`SCM_RIGHTS`][socket.h],
+    * `cmsg_len` set to `CMSG_LEN(sizeof(int))`, and
+    * `*CMSG_DATA(cmsg)` set to the pseudoterminal master file descriptor.
+
+##### Responses
+
+All responses MUST contain a **`type`** property whose value MUST one of the following strings:
+
+* `success`, if the request was successfully processed.
+* `error`, if the request was not successfully processed.
+
+In addition, responses MAY contain any of the following properties:
+
+* **`message`** (string, OPTIONAL) A short phrase describing the response.
+
+#### Example
+
+```
+# in a bundle directory with a process that echos "hello" and exits 42
+$ test -t 1 && echo 'stdout is a terminal'
+stdout is a terminal
+$ funC create hello-1
+$ echo $?
+0
+$ funC start hello-1
+hello
+$ echo $?
+0
+$ block-on-exit-and-collect-exit-code hello-1
+$ echo $?
+42
+$ funC delete hello-1
+$ echo $?
+0
+```
+
+`hello` shows up in the terminal after `start`, because `start` happens to be called from the same shell session (using the same terminal) as `create`, and the container process inherited its standard streams from `create`.
+If you call `start` from a shell using a separate terminal, the container output would still have appeared in the `create` terminal.
```

Comment:
> That's the expected behavior, and not a kernel bug? Can you link me to the LXC discussion so I can do more background reading without pestering you here?

I will find a link -- there's a commit in LXC that references this.

> I think “gets the terminal for stdio” is pretty common default behavior. And it's easy to set your stdio up explicitly when you don't want that.

I don't think it's common behaviour for containers (whatever "common behaviour" means in this context, considering that container runtimes aren't very common). Most importantly, it's odd for two reasons:

1. `runc create` detaches from the terminal. Now, this isn't _hugely_ odd with most programs (essentially all programs detached with `&` do this, though to be fair it's actuall a shell thing and not a `--detach` thing -- most programs with a `--detach` flag don't spam the terminal's stdio).

2. We're creating **containers**, which are meant to be isolated from the host. When I am in an interactive `runc run` session I expect that my IO is being **relayed** to the container (which is actually what's happening). But when that is detached, I would expect that since there's nothing to "relay" the IO that the container **is not** connected to my terminal. But it is connected to my terminal.

So IMO this is not intuitive or expected behaviour by a user (especially in the `runc create` case).